### PR TITLE
Added notifications for Confirmed, Error and Dropped status changes

### PR DESCRIPTION
**Pull request by spylogsster**
Resolves https://github.com/brave/brave-browser/issues/19885

Added **WalletNotificationService** to show notifications about transactions on UI level.

https://user-images.githubusercontent.com/2965009/168286273-05c025bc-db8f-4344-b656-784ca1c3e77b.mp4

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:

- Make transaction and check notification for Confirmed and Error statuses

### CODE REVIEW

File: browser/brave_wallet/wallet_notification_service.h [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+namespace brave_wallet {
+
+class TxService;
+
+class WalletNotificationService : public KeyedService,
+                                  public mojom::TxServiceObserver {
+ public:
+  explicit WalletNotificationService(content::BrowserContext* context,
+                                     TxService* tx_state_manager);
```
Comment (yrliou) [inline browser/brave_wallet/wallet_notification_service.h:29]: s/tx_state_manager/tx_service

File: browser/brave_wallet/wallet_notification_service.cc [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+
+}  // namespace
+
+namespace brave_wallet {
+
+WalletNotificationService::WalletNotificationService(
+    content::BrowserContext* context,
+    TxService* tx_service)
+    : tx_service_(tx_service), context_(context) {
+  tx_service_->AddObserver(tx_observer_receiver_.BindNewPipeAndPassRemote());
```
Comment (yrliou) [inline browser/brave_wallet/wallet_notification_service.cc:64]: This raw_ptr doesn't seem to be needed to be stored if all we need is the AddObserver call here.

File: browser/brave_wallet/wallet_notification_service.h [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+#ifndef BRAVE_BROWSER_BRAVE_WALLET_WALLET_NOTIFICATION_SERVICE_H_
+#define BRAVE_BROWSER_BRAVE_WALLET_WALLET_NOTIFICATION_SERVICE_H_
+
+#include <string>
+
+#include "base/memory/raw_ptr.h"
+#include "brave/components/brave_wallet/common/brave_wallet.mojom.h"
+#include "components/keyed_service/core/keyed_service.h"
+#include "mojo/public/cpp/bindings/pending_remote.h"
+#include "mojo/public/cpp/bindings/remote.h"
```
Comment (yrliou) [inline browser/brave_wallet/wallet_notification_service.h:15]: These two seems not needed in this header file, and the include for mojo::Receiver is missing.

File: browser/brave_wallet/wallet_notification_service.cc [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+}
+
+WalletNotificationService::~WalletNotificationService() {}
+
+bool WalletNotificationService::ShouldDisplayUserNotification(
+    mojom::TransactionStatus status) {
```
Comment (yrliou) [inline browser/brave_wallet/wallet_notification_service.cc:70]: I think we should probably display the notification for Dropped too?

File: browser/brave_wallet/brave_wallet_tab_helper.cc [hunk 5]
```diff
@@ -23,7 +24,10 @@
 namespace brave_wallet {
 
 BraveWalletTabHelper::BraveWalletTabHelper(content::WebContents* web_contents)
-    : content::WebContentsUserData<BraveWalletTabHelper>(*web_contents) {}
+    : content::WebContentsUserData<BraveWalletTabHelper>(*web_contents) {
+  brave_wallet::WalletNotificationServiceFactory::GetServiceForContext(
+      web_contents->GetBrowserContext());
```
Comment (yrliou) [inline browser/brave_wallet/brave_wallet_tab_helper.cc:29]: I'm a bit confused about this. Could you explain what's this for? Triggering service creation? How is it related to the tab helper?

File: browser/brave_wallet/brave_wallet_provider_delegate_impl_helper.cc [hunk 6]
```diff
@@ -14,7 +14,7 @@
 
 namespace {
 
-base::OnceCallback<void()> g_NewSetupNeededForTestingCallback;
+absl::optional<bool> g_NewSetupNeededForTesting;
```
[the remark this answers is not in the document]
Comment (spylogsster) [reply browser/brave_wallet/brave_wallet_provider_delegate_impl_helper.cc:17]: there is condition of existance on 39th line, replaced by optional to keep same logic

File: browser/brave_wallet/BUILD.gn [hunk 7]
```diff
@@ -36,7 +38,6 @@ source_set("brave_wallet") {
     "//services/network/public/cpp",
     "//third_party/abseil-cpp:absl",
   ]
-
```
[the remark this answers is not in the document]
Comment (spylogsster) [reply browser/brave_wallet/BUILD.gn]: done

File: browser/brave_wallet/wallet_notification_service.cc (hunk 4 above)
Comment (spylogsster) [reply browser/brave_wallet/wallet_notification_service.cc:84]: added

File: browser/brave_wallet/brave_wallet_tab_helper.cc (hunk 5 above)
Comment (spylogsster) [reply browser/brave_wallet/brave_wallet_tab_helper.cc:29]: yes, this is only to trigger service creation, it needs to be done on UI level for browser context

File: browser/brave_wallet/wallet_notification_service.h (hunk 3 above)
Comment (spylogsster) [reply browser/brave_wallet/wallet_notification_service.h]: done

File: browser/brave_wallet/wallet_notification_service.cc (hunk 2 above)
Comment (spylogsster) [reply browser/brave_wallet/wallet_notification_service.cc]: done

File: browser/brave_wallet/wallet_notification_service.h (hunk 1 above)
Comment (spylogsster) [reply browser/brave_wallet/wallet_notification_service.h]: done

File: test/BUILD.gn [hunk 8]
```diff
@@ -88,6 +88,7 @@ test("brave_unit_tests") {
     "../../components/domain_reliability/test_util.h",
     "//brave/browser/brave_content_browser_client_unittest.cc",
     "//brave/browser/brave_resources_util_unittest.cc",
+    "//brave/browser/brave_wallet/wallet_notification_service_unittest.cc",
```
[the remark this answers is not in the document]
Comment (spylogsster) [reply test/BUILD.gn:91]: I couldn't find any mention of unittests in underlaying `sources.gni` files, seems this is only way now to add test for this on UI level

File: browser/sources.gni [hunk 9]
```diff
@@ -102,8 +103,8 @@ brave_chrome_browser_deps = [
   "//brave/browser:browser_process",
   "//brave/browser/brave_federated",
   "//brave/browser/brave_wallet",
+  "//brave/browser/brave_wallet",
```
[the remark this answers is not in the document]
Comment (spylogsster) [reply browser/sources.gni]: done

File: browser/brave_wallet/sources.gni [hunk 10]
```diff
@@ -0,0 +1,27 @@
+brave_browser_brave_wallet_sources = [
```
[the remark this answers is not in the document]
Comment (spylogsster) [reply browser/brave_wallet/sources.gni:6]: done

File: browser/brave_wallet/brave_wallet_provider_delegate_impl_helper.cc (hunk 6 above)
[the remark this answers is not in the document]
Comment (yrliou) [reply browser/brave_wallet/brave_wallet_provider_delegate_impl_helper.cc:17]: Seems unnecessary, I think a boolean value initialized as false and just check the boolean value is what we needed.

File: browser/brave_wallet/sources.gni [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this file,
+# You can obtain one at http://mozilla.org/MPL/2.0/.
+
+brave_browser_brave_wallet_sources = [
+  "//brave/browser/brave_wallet/brave_wallet_tab_helper.cc",
+  "//brave/browser/brave_wallet/brave_wallet_tab_helper.h",
+]
+
+if (!is_android) {
```
Comment (bridiver) [inline browser/brave_wallet/sources.gni:11]: do not create new sources.gni files. There is no reason for new code to have circular dependency issues

File: browser/brave_wallet/wallet_notification_service.cc [hunk 12]
```diff
@@ -0,0 +1,109 @@
...
+  NotificationDisplayServiceFactory::GetForProfile(profile)->Display(
+      NotificationHandler::Type::SEND_TAB_TO_SELF, *notification, nullptr);
+}
+
+}  // namespace
+
+namespace brave_wallet {
+
+WalletNotificationService::WalletNotificationService(
+    content::BrowserContext* context,
```
Comment (bridiver) [inline browser/brave_wallet/wallet_notification_service.cc:75]: do not pass in the BrowserContext, pass in NotificationDisplayService from WalletNotificationServiceFactory

File: browser/brave_wallet/brave_wallet_provider_delegate_impl_helper.cc (hunk 6 above)
[the remark this answers is not in the document]
Comment (bridiver) [reply browser/brave_wallet/brave_wallet_provider_delegate_impl_helper.cc:17]: this variable name also needs to be underscore

File: browser/brave_wallet/sources.gni [hunk 13]
```diff
@@ -0,0 +1,32 @@
+# Copyright (c) 2022 The Brave Authors. All rights reserved.
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this file,
+# You can obtain one at http://mozilla.org/MPL/2.0/.
+
+brave_browser_brave_wallet_sources = [
```
Comment (bridiver) [inline browser/brave_wallet/sources.gni:6]: so based on dm discussion brave_wallet_tab_helper shouldn't need to be in sources.gni anymore so I would move it back to BUILD.gn and then create a subdir for notifications so we don't have sources.gni and BUILD.gn in the same directory

Comment (spylogsster) [reply browser/brave_wallet/sources.gni]: done

File: browser/brave_wallet/wallet_notification_service.cc (hunk 12 above)
Comment (spylogsster) [reply browser/brave_wallet/wallet_notification_service.cc]: I need BrowserContext inside `WalletNotificationService`

File: test/BUILD.gn [hunk 14]
```diff
@@ -312,6 +313,7 @@ test("brave_unit_tests") {
     sources += [
       "../utility/importer/chrome_importer_unittest.cc",
       "//brave/app/brave_command_line_helper_unittest.cc",
+      "//brave/browser/brave_wallet/notifications/wallet_notification_service_unittest.cc",
```
[the remark this answers is not in the document]
Comment (spylogsster) [reply test/BUILD.gn]: done

**Review by yrliou — APPROVED** [review]
LGTM

**Pull request merged**